ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/validation/RequestValidator.java

```diff
@@ -0,0 +1,145 @@
+import io.confluent.ksql.util.KsqlException;
+import io.confluent.ksql.util.KsqlStatementException;
+import io.confluent.ksql.util.StatementWithSchema;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.function.Supplier;
+
+/**
+ * Wraps an execution context and information about how to validate statements
+ * in a way that can generate a validated checksum against a KSQL query without
+ * worrying about races that may occur.
+ */
+public class RequestValidator {
+
+  private final Map<Class<? extends Statement>, StatementValidator> customValidators;
+  private final Set<Class<? extends Statement>> optionalToValidate;
+  private final Supplier<EngineSnapshot> snapshotSupplier;
+  private final SchemaRegistryClient registryClient;
+  private final KsqlConfig ksqlConfig;
+
+  /**
+   * @param customValidators        a map describing how to validate each statement of type
+   * @param optionalToValidate      these statements will not cause validation to fail if there
+   *                                is no registered validator
+   * @param snapshotSupplier        supplies a snapshot of the current execution state, the
+   *                                snapshot returned will be owned by this class and changes
+   *                                to the snapshot should not affect the source and vice versa
+   * @param registryClient          a client to use for inferring schema for statements
+   * @param ksqlConfig              the {@link KsqlConfig} to validate against
+   */
+  public RequestValidator(
+      final Map<Class<? extends Statement>, StatementValidator> customValidators,
+      final Set<Class<? extends Statement>> optionalToValidate,
+      final Supplier<EngineSnapshot> snapshotSupplier,
+      final SchemaRegistryClient registryClient,
+      final KsqlConfig ksqlConfig
+  ) {
+    this.customValidators = customValidators;
+    this.optionalToValidate = optionalToValidate;
+    this.snapshotSupplier = snapshotSupplier;
+    this.registryClient = registryClient;
+    this.ksqlConfig = ksqlConfig;
+  }
+
+  /**
+   * Validates the messages against a snapshot in time of the KSQL engine.
+   *
+   * @param statements          the list of statements to validate
+   * @param propertyOverrides   a map of properties to override for this validation
+   * @param sql                 the sql that generated the list of statements, used for
+   *                            generating more useful debugging information
+   *
+   * @throws KsqlException if any of the statements cannot be validated, or the number
+   *                       of requested statements would cause the execution context
+   *                       to exceed the number of persistent queries that it was configured
+   *                       to support
+   */
+  public void validate(
+      final List<ParsedStatement> statements,
+      final Map<String, Object> propertyOverrides,
+      final String sql
```

Comment:
Feels a shame to have to the full sql all the way down.  You're basically passing passing the string all the way down the call stack, only to throw it back up to something that already has it.

Personally, I'd drop the `sql` parameter and just ensure `KsqlResource` includes it when handling the capacity exception.

Reply:
I agree, but I'd like to do this in another PR since this one is already getting unwieldily